Traefik logging and grafana-dashboard integrations

PietroPasotti opened the pull request:
This PR adds loki-logging and grafana-dashboarding integrations to the traefik charm.

## Testing Instructions
deploy cos-lite, then refresh the traefik charm with one built from these sources. Visit the grafana frontend to see the traefik dashboard with the logs.

![image](https://user-images.githubusercontent.com/6230162/211538053-ea03b5a9-dfc6-4f43-b0c2-80f197993869.png)

## Release Notes
- Added loki-logging and grafana-dashboarding integrations.

Abuelodelanada commented:
> Should we add a loki alert? Seems like a good start could be a general purpose alert for when there's an ERROR level entry in the logs.

It's in another PR: https://github.com/canonical/traefik-k8s-operator/pull/120

PietroPasotti commented:
> * Drop down "Juju Model" is title-cased unlike the others.
Fixed

> * Drop downs "Job" and "Interval": what are they used for?
Unsure, they came with the dashboard (https://grafana.com/grafana/dashboards/4475-traefik/)

> * Pie chart's title is missing
Fixed

> * Average response time: should we add to the title the time range? E.g. "Average response time (last 5 min)". Unless it's averaging across something else than time?
Looking at the formula it looks like it's time. And the graph is showing the last hour, so I added that. But perhaps it was meant to be used together with that 'interval' dropdown somehow?

> * Used sockets: should we set the type of the y axis to integers?
As soon as there's data the values automatically go to integers, but I'll try to hardcode it

sed-i commented:
> Drop downs "Job" and "Interval": what are they used for?
> Unsure, they came with the dashboard (https://grafana.com/grafana/dashboards/4475-traefik/)

:shrug: Any chance it's redundant given juju topology?

> * Looking at the formula it looks like it's time. And the graph is showing the last hour, so I added that. But perhaps it was meant to be used together with that 'interval' dropdown somehow?

Maybe you're right: the expression doesn't specify duration:
```
"expr": "sum(traefik_backend_request_duration_seconds_sum{protocol=~\"$protocol\"}) / sum(traefik_entrypoint_requests_total{protocol=~\"$protocol\"}) * 1000"
```
but it also doesn't use `$__interval`. Interesting. Perhaps `"intervalFactor": 2` has anything to do with it?

lucabello commented on src/grafana_dashboards/traefik.json.tmpl:
```diff
@@ -0,0 +1,1454 @@
+        "includeAll": false,
+        "label": "juju_unit",
+        "multi": false,
+        "name": "juju_unit",
```
Minor pet peeve, I would change this variable display name to `Juju unit` to match the style of `Juju application` and `Juju model`
```suggestion
        "name": "Juju unit",
```

lucabello commented on src/grafana_dashboards/traefik.json.tmpl:
```diff
@@ -0,0 +1,1454 @@
+      ],
+      "thresholds": [],
+      "timeRegions": [],
+      "title": "ENTRYPOINT - Open Connections",
```
Another minor pet peeve, me not liking full caps
```suggestion
      "title": "Entrypoint - Open Connections",
```

lucabello commented on src/grafana_dashboards/traefik.json.tmpl:
```diff
@@ -0,0 +1,1454 @@
+      ],
+      "thresholds": [],
+      "timeRegions": [],
+      "title": "BACKEND - Open Connections",
```
Another minor pet peeve
```suggestion
      "title": "Backend - Open Connections",
```

lucabello reviewed (approved):
I have just a couple minor aesthetic remarks, feel free to ignore them :)

rbarry82 commented on src/grafana_dashboards/traefik.json.tmpl:
```diff
@@ -0,0 +1,1454 @@
+        "type": "datasource"
+      },
+      {
+        "allValue": "traefik",
```
Technically speaking, all of these dropdowns aren't necessary and will be injected by the dashboard library, but they're not hurting anything here either.

dstathis commented on src/charm.py:
```diff
@@ -866,7 +881,8 @@ def _restart_traefik(self):
                     "override": "replace",
                     "summary": "Traefik",
-                    "command": BIN_PATH,
+                    # trick to drop the logs to a file but also keep them available in the pod logs
+                    "command": '/bin/sh -c "{} | tee {}"'.format(BIN_PATH, self._log_path),
```
+1 This is great

PietroPasotti replied:
I don't remember who I ripped it off of; it's not flour from my own bag (*italian idiom)
